### PR TITLE
Documentation tests

### PR DESCRIPTION
**Pull request by AdrianSosic**
Adds tests for the code blocks in the README and our user guides. Also, fixes one bug that got unveiled.

### CODE REVIEW

File: docs/userguide/constraints.md [hunk 1]
```diff
@@ -99,7 +99,7 @@ from baybe.constraints import ThresholdCondition
 
 ThresholdCondition( # will select all values above 150
     threshold = 150, 
-    operator = ">",
+    operator = "=",
```
Comment (AdrianSosic) [inline docs/userguide/constraints.md:102]: @Scienfitz: I guess this is correct now?

File: tests/docs/test_docs.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+from pathlib import Path
+
+import pytest
+
+from .utils import extract_code_blocks
+
+
+def test_readme():
+    """The blocks in the README become a valid python script when concatenated."""
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/docs/test_docs.py:11]: I know what you mean and also thought about whether this is desired or not. But let me try to convince you:
- What is the minimal alternative? --> We should **at least** require that each code block itself is executable, without throwing errors and code blocks should not necessarily need to know that other blocks exist, right? Well, if each code block is itself executable, then they **are** also executable also as a script. That is, local executability IMPLIES global executability. So really, we really lose nothing and are gaining only.
- In fact, there might be cases (and currently README **is** designed in this way) that a script can only work if previous code blocks are executed first. So this falls in the more general class which would not work with block based tests.
- Even IF there are future cases where we need an even more flexible approach (which I really can't imagine right now), then we can still adjust. Nothing is set in stone here. I simply coded it in the most general way for NOW. 

Does it make sense? Or do you have another suggestion?

[the remark this answers is not in the document]
Comment (AVHopp) [reply tests/docs/test_docs.py:11]: For me, the examples are the place to go if you want to see the actual code and fully working examples. If we force everything to already be fully functional code including explanation - what is the difference between specifically user guides and examples? Thus, the user guides should just *explain* stuff and include the absolute *minimal* amount of code necessary for the purpose of *explanation*, not for *executing*.

[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/docs/test_docs.py:11]: OK, I can live with your point that code snippets do not necessarily have to be self-contained and they might depend on other stuff that is NOT shown in the code block (which is BTW exactly my point for the README above: it uses stuff defined earlier). But even then, two remarks:
* We can still easily turn it into an intact script by having a header setting in the guide that sets the context. For example, for the search space, something like _"In the following examples, we assume that `parameters` holds an arbitrary number of parameters. For example:"_
   ```python
   <example set up code here>
   ```
   That would make it executable again AND more explicit what are the assumptions about the elements for the shown examples.
* I am completely happy to switch to a different testing approach, but **not** testing the docs examples is really a no-go for me. Reason: there is zero (and I really mean zero) chance here that the doc will not go out of sync with our code. It's completely impossible to keep this intact manually while we are still strongly changing interfaces and structures in the code. The bug my tests found is already a first (and very minor) example of this, and we have already way too many parts in the examples that either have high risk of getting out of sync, are already out of sync, or required too much manual effort to resync during previous PRs. If we don't take care of these things soon, I predict we'll spent much of our time on trying to keep the docs up with our code.

So long story short 😄: give me a better *testable* alternative. We can absolutely postpone this to a dev meeting, though 👍🏼

File: docs/userguide/constraints.md (hunk 1 above)
Comment (Scienfitz) [reply docs/userguide/constraints.md:102]: took me some time to understand what was even going on here, i think you fixed the wrong thing, the condition here does not fit "will select values above 150" as written in the comment anymore. So imo either i) keep the original simply removing the tolerance or ii) take what you've done here but rewrite the accompanying comment

File: tests/docs/test_docs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AVHopp) [reply tests/docs/test_docs.py:11]: After having thought about this and after having re-written the campaign user guide I have to admit: *I was wrong and you were right* (enjoy this statement as much as you want 😉 )
In my opinion, the best alternative is to have the code blocks as self-contained as possible (e.g. having imports only in the cell for which they are necessary), while allowing to re-use stuff from earlier cells - which should be testable when we just dump everything together and run it right?
So feel free to write a comment like "TOLD YOU SO" and resolve this conversation 😄

[the remark this answers is not in the document]
Comment (AVHopp) [reply tests/docs/test_docs.py:11]: Once you've done that I will change my review to approve ;)

[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/docs/test_docs.py:11]: Hey @AVHopp, no need for shame here, I'm always happy about critical feedback and being challenged :D And I also have good news for you: I thought about it, too, and decided there also a middle ground! I still think that the code **must** be testable, but in case we really get to the point where we don't want to declare all necessarily variables in earlier code blocks (or it simply would not fit to the text around it), we can actually also fill the missing gaps on the **test side**, using a fixture that provides "common defaults".

To explain what I mean: Say you are talking about campaigns in the guide and want to show a piece of code where some arbitrary parameters are required but you really don't want to explicitly mention them because the guide is not about parameter. For instance, you have:
```pythton
campaign = Campaign(searchspace, objective, strategy)
```
and we somewhere need to have parameters defined in order to have a searchspace object to start with, but you don't want to explicitly show them because their particular values don't matter for the purpose of your guide. 

Then, what we can do is have a generic set of common objects provided through a fixture that are loaded into the  namespace before the test blocks are executed! Of course, the code in the test cells then needs to be written in a way that it can cope with "arbitrary default objects", but this is completely reasonable assumption. Because if you require something very special, then you would explicitly add this to your code.

What do you think about it? If you want, I can set up the necessary architecture already in this PR and we fill the "default objects" on the fly, as soon as they are needed. Or we do so once we really encounter the first use case.

[the remark this answers is not in the document]
Comment (Scienfitz) [reply tests/docs/test_docs.py:11]: why so complicated? is it not possible to have a simple flag or comment that would turn the check off for that cell? imo that would be most pragmatical. Overall I agree we need the checks but to avoid spamming the userguides with boilerplate one should simply be able to turn that off.

Assuming global execution from local execution imo is not right, imagine a situation with cells A,B...Z. Z depends on A. But some cell say C makes a change to A that is intended as a demonstration and not in the global context of all cells (in fact it could break Z and others etc)

[the remark this answers is not in the document]
Comment (AVHopp) [reply tests/docs/test_docs.py:11]: In principle, this sounds good and interesting 😄  How much code would this require us to include into the user guides/README? If this requires a lot of or hard to understand additional code, we might confuse the readers.

You could have a look at the campaign user guide PR. In my opinion, the code that is not related to the campaigns themselves is not too much, and having the cells self-contained does work. But I think we should discuss this internally, as I'm also interested about @Scienfitz opinion here. So I wouldn't suggest already setting things up before we have his opinion.

For me, the only important things are 
- code should be testable and tested
- additional code that is not related to the functionalities of BayBE in general should be minimal and not confuse the reader

File: docs/userguide/constraints.md (hunk 1 above)
Comment (AdrianSosic) [reply docs/userguide/constraints.md]: Ah, I see, fixed it. This now resolves #68. I think the fact that this got noticed so quickly by users proves that we need automatic tests 😄

**Comment by AdrianSosic** [comment]
@Scienfitz @AVHopp, two accepts and I will merge 🚀

File: tests/docs/test_docs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/docs/test_docs.py:11]: Agreed on a strategy in the meeting: will take as is for now and adjust once necessary.

**Review by Scienfitz — APPROVED** [review]

**Review by AVHopp — APPROVED** [review]
